_doc-build-linux/action.yml

```diff
@@ -185,56 +185,112 @@ runs:
         sudo apt-get update
         sudo apt-get install -y poppler-utils
 
-    - name: "Check if requirements.txt file exists"
+    # ------------------------------------------------------------------------
+
+    - uses: ansys/actions/_logging@main
+      with:
+        level: "INFO"
+        message: >
+          Determine context.
+
+    - name: "Determine GitHub environment variables"
       shell: bash
       run: |
-        echo "EXISTS_DOC_REQUIREMENTS=$(if [ -f ${{ inputs.requirements-file }} ]; then echo 'true'; else echo 'false'; fi)" >> $GITHUB_ENV
+        if [[ -f "pyproject.toml" ]] && grep -q 'build-backend = "poetry\.core\.masonry\.api"' "pyproject.toml"; then
+          echo "BUILD_BACKEND=$(echo 'poetry')" >> $GITHUB_ENV
+          echo "SPHINX_BUILD_MAKE=$(echo 'poetry run -- make')" >> $GITHUB_ENV
+        else
+          echo "BUILD_BACKEND=$(echo 'pip')" >> $GITHUB_ENV
+          echo "SPHINX_BUILD_MAKE=$(echo 'make')" >> $GITHUB_ENV
+        fi
+
+    - uses: ansys/actions/_logging@main
+      with:
+        level: "INFO"
+        message: >
+          Build backend: ${{ env.BUILD_BACKEND }}
+          Sphinx build make: ${{ env.SPHINX_BUILD_MAKE }}
+
+    # ------------------------------------------------------------------------
+
+    - uses: ansys/actions/_logging@main
+      with:
+        level: "INFO"
+        message: >
+          Set up python to build the documentation.
 
-    - name: "Print previous output"
+    - name: "Set poetry environment variable(s) (if required)"
       shell: bash
+      if: env.BUILD_BACKEND == 'poetry'
       run: |
-        echo "Output was found ${{ env.EXISTS_DOC_REQUIREMENTS }}"
+        # For projects using poetry, do not use a virtual environment.
+        # Poetry uses virtual environments to install its dependencies, but this might
+        # lead to problems if it is not activated prior to executing poetry commands.
+        # Store POETRY_VIRTUALENVS_CREATE=false in the GitHub environment to prevent
+        # poetry from creating a virtual environment.
+        #
+        echo "POETRY_VIRTUALENVS_CREATE=false" >> $GITHUB_ENV
 
     - name: "Update pip"
       shell: bash
-      run: python -m pip install -U pip
+      run: |
+        source .venv/bin/activate
+        python -m pip install -U pip
+
+    - name: "Check if requirements.txt file exists"
+      shell: bash
+      run: |
+        echo "EXISTS_DOC_REQUIREMENTS=$(if [ -f ${{ inputs.requirements-file }} ]; then echo 'true'; else echo 'false'; fi)" >> $GITHUB_ENV
 
     - name: "Install documentation dependencies from requirements file"
       shell: bash
       if: env.EXISTS_DOC_REQUIREMENTS == 'true'
       run: |
+        source .venv/bin/activate
         python -m pip install -r ${{ inputs.requirements-file }}
 
+    - name: "Install poetry (if required)"
+      shell: bash
+      if: env.BUILD_BACKEND == 'poetry'
+      run: |
+        source .venv/bin/activate
+        python -m pip install poetry
```

Comment:
Installing poetry into the same virtual environment that the project is installed into isn't recommended. If poetry and the running package have shared dependencies with different versions, poetry tries to upgrade/downgrade its dependencies while it's running, which often fails, especially on Windows.

The workaround we use is to use `pipx` to install poetry. Something like:

```yaml
      - name: "Add pipx/bin dir to path"
        run: echo "${{ runner.temp }}/pipx/bin" >> $GITHUB_PATH

      - name: "Install poetry and create empty virtual environment"
        run: |
          python -m pip install pipx --user
          python -m pipx install poetry
        env:
          PIPX_BIN_DIR: ${{ runner.temp }}/pipx/bin
          PIPX_HOME : ${{ runner.temp }}/pipx/home
```

Installing pipx into the `${{ runner.temp }}` directory means that the pipx installation will automatically be cleaned up between jobs on a VM. Adding the bin directory to the path first means pipx shouldn't complain that the poetry executable isn't on path.

This way, poetry is kept separate to the project, and they are free to use different versions of the same shared dependencies.

Reply:
This _does_ mean that pipx is installed outside of any virtual environment, but I think that's only an issue if you really care about using a specific version of pipx between different jobs, which I imagine is a pretty rare requirement.